lib/rust/performance-logger/src/lib.rs

```diff
@@ -0,0 +1,395 @@
+
+use enso_prelude::*;
+
+use enso_prelude::fmt::Formatter;
+use enso_web::performance;
+use ordered_float::OrderedFloat;
+use serde::Deserialize;
+use serde::Serialize;
+use web_sys::PerformanceEntry;
+
+
+
+// ========================================
+// === Compile time filtering constants ===
+// ========================================
+
+macro_rules! define_profiling_toggle {
+    ($log_level_name_upper:ident, $log_level_name:ident) => {
+        paste::paste! {
+            #[doc = "Defines whether the log level `" $log_level_name "` should be used."]
+            #[cfg(feature = "enable-" $log_level_name "-profiling")]
+            pub const [< ENABLE_ $log_level_name_upper _PROFILING >]: bool = true;
+            #[doc = "Defines whether the log level `" $log_level_name "` should be used."]
+            #[cfg(not(feature = "enable-" $log_level_name  "-profiling"))]
+            pub const [< ENABLE_ $log_level_name_upper _PROFILING >]: bool = false;
+        }
+    };
+}
+
+define_profiling_toggle!(SECTION, section);
+define_profiling_toggle!(TASK, task);
+define_profiling_toggle!(DETAIL, detail);
+define_profiling_toggle!(DEBUG, debug);
+
+
+
+// ====================
+// === Measurements ===
+// ====================
+
+
+/// A single interval measurement.
+#[derive(Debug, Clone)]
+#[allow(missing_docs)]
+pub struct Measurement {
+    pub name:       String,
+    pub start_time: f64,
+    pub duration:   f64,
+    pub log_level:  LogLevel,
+}
+
+impl Measurement {
+    /// Return the timestamp of the interval's end.
+    pub fn end_time(&self) -> f64 {
+        self.start_time + self.duration
+    }
+}
+
+impl Display for Measurement {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        f.write_str(&format!(
+            "[{:.2}, {:.2}] ({:.2}) {}",
+            self.start_time,
+            self.end_time(),
+            self.duration,
+            self.name
+        ))
+    }
+}
+
+/// Error that can occur when converting between a `PerformanceEntry` and a `Measurement`.
+#[derive(Debug, Clone)]
+pub enum MeasurementConversionError {
+    /// The name of the `PerformanceEntry` was not in a format that could be decoded to a
+    /// `Measurement`.
+    InvalidFormatting,
+    /// The log level encoded in the `PerformanceEntry`s name is not a valid log level.
+    InvalidLogLevel,
+}
+
+impl TryFrom<PerformanceEntry> for Measurement {
+    type Error = MeasurementConversionError;
+
+    fn try_from(measure: PerformanceEntry) -> Result<Self, Self::Error> {
+        use MeasurementConversionError::*;
+
+        let start_time = measure.start_time();
+        let duration = measure.duration();
+        let name_js = measure.name();
+        let name_parts: Vec<_> = name_js.split(MESSAGE_DELIMITER).collect();
+        if name_parts.len() < 2 {
+            return Err(InvalidFormatting); //FIXME
+        }
+        let name = name_parts[1].to_string();
+        let log_level_name = name_parts[0].to_string();
```

Comment:
Lines that can crash during runtime. Previous check to <2 is not enough. During refactoring it can break.

Reply:
Also fixed with previous fix.